### ISSUE TITLE
ToC cell + LaTeX in a heading + HTML export

### ISSUE THREAD

**Issue by PeterMinin** [ipython-contrib/IPython-notebook-extensions#635]
This combination (with the table of contents generated with ToC2) leads to a broken html page: the LaTeX part of the heading "runs away" from its place and floats next to the table of contents. See the attachment for an example.
[ToC cell + LaTeX in a heading + HTML export.zip](https://github.com/ipython-contrib/IPython-notebook-extensions/files/322026/ToC.cell.LaTeX.in.a.heading.HTML.export.zip)

**Comment by jfbercher** [comment]
If you haven't done it yet, please update to the last version. 
Then try 
```jupyter nbconvert FILE.ipynb --template toc2```
(see docs in README)
in your case
```
jupyter nbconvert "ToC cell + LaTeX in a heading + HTML export.ipynb" --template toc2
```
It should give you the attached file. 

[ToC cell + LaTeX in a heading + HTML export.html.zip](https://github.com/ipython-contrib/IPython-notebook-extensions/files/322167/ToC.cell.LaTeX.in.a.heading.HTML.export.html.zip)

**Comment by jfbercher** [comment]
@PeterMinin Any progress regarding this issue? Does it works as intended now?  

It is now also possible to export to html with toc embedded using `jupyter nbconvert --to html_toc FILE.ipynb`

**Comment by PeterMinin** [comment]
I'm really sorry about leaving it like this. At the time, I was indeed on an older version. After updating I still had some problems, partially caused by my setup, so I wanted to figure everything out there before I could say anything definite. Should've known it would take an infinite time... :)

Right now, testing on a different system in a fresh conda env with the latest master version, things are better, but there are still problems. I found an apparently reproducible case.

Say, I have 3 headings with LaTeX. If I start editing all of them and then format them at the same time using **Cell -> Run all**, I get a good output in HTML (right), but the notebook's ToC sidebar displays some headings unformatted (left).

<a href="https://cloud.githubusercontent.com/assets/2499965/20860912/5df863a6-b994-11e6-80a3-6eedc47740ce.png"><img src="https://cloud.githubusercontent.com/assets/2499965/20860912/5df863a6-b994-11e6-80a3-6eedc47740ce.png" align="left" width="300" ></a>

<a href="https://cloud.githubusercontent.com/assets/2499965/20860916/62adad16-b994-11e6-831f-8651c1662ed1.png"><img src="https://cloud.githubusercontent.com/assets/2499965/20860916/62adad16-b994-11e6-831f-8651c1662ed1.png" width="300" ></a>

But if I start editing the headings one by one, only the last one is displayed correctly in HTML.

<a href="https://cloud.githubusercontent.com/assets/2499965/20860910/55d6e788-b994-11e6-8550-6657d83fe7dc.png"><img src="https://cloud.githubusercontent.com/assets/2499965/20860910/55d6e788-b994-11e6-8550-6657d83fe7dc.png" align="left" width="300" ></a>

<a href="https://cloud.githubusercontent.com/assets/2499965/20860918/6816f334-b994-11e6-9f85-3e5483cc762e.png"><img src="https://cloud.githubusercontent.com/assets/2499965/20860918/6816f334-b994-11e6-9f85-3e5483cc762e.png" width="300" ></a>

Before the Nov 16 changes (the 0.2.2 release on conda-forge still doesn't have them for some reason) I was also sometimes getting a red "[Math Processing Error]" instead of LaTeX, which is also how my old broken HTML is now displayed.

**Comment by jcb91** [comment]
> the 0.2.2 release on conda-forge still doesn't have them for some reason
I guess you say this because 0.2.2 only got released today on conda-forge? I'm afraid the release schedule is quite a bit behind master at the moment!

**Comment by PeterMinin** [comment]
Not really, yesterday I noticed it was on 0.2.1, so I checked again today and it downloaded 0.2.2.

**Comment by PeterMinin** [comment]
Apparently I didn't reload something. I checked again, and the changes are there. Sorry for the confusion.

**Comment by jcb91** [comment]
> I checked again today and it downloaded 0.2.2

Yeah, I only just got around to pushing 0.2.2 to conda-forge today. The forge recipe pulls from pypi.io rather than pypi.org (at the request of the forge folks), and since pypi.io gets source at a delay from pypi.python.org (variable, from minutes to days, I don't know why), the conda recipe updates lag behind the pip uploads. This release was also complicated by the fact I needed to rerender the recipe for a newer conda-smithy version, and that got delayed by long delays in Travis OSX build infrastructure over the last couple of days.

If you want a bleeding-edge version through conda, you can build the recipe in this repo using [the method in our condarecipe tox env](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/3c290a2dd1acd41ab4bd37a0f09a6d9380c6cda1/tox.ini#L157-L158), which is used for the [travis conda-recipe build](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/3c290a2dd1acd41ab4bd37a0f09a6d9380c6cda1/.travis.yml#L27-L29):

    conda build --channel conda-forge conda.recipe
    conda install --channel conda-forge --use-local jupyter_contrib_nbextensions

or the similar

    conda build --channel conda-forge conda.recipe
    conda update --channel conda-forge --use-local jupyter_contrib_nbextensions

to update. Version numbers should be compatible with the main conda-forge releases.

**Comment by jfbercher** [comment]
Concerning the "$\LaTeX$ issue mentioned above, I have been able to reproduce transiently (but only transiently). I think that this might be a mathjax related issue, as also shown by the [Math Processing Error] which is a mathjax error message. 
Perhaps of course that it is not standard/supported to include mathjax maths in links and that we should filter out maths from toc headers? [though I remember that we found useful [to enable it](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/pull/394#r44070908)... ]

**Comment by jcb91** [comment]
> Perhaps of course that it is not standard/supported to include mathjax maths in links

I doubt it's unsupported, after all, html links should be able to contain almost any other tag, no? Certainly imd, div, span and the like which I'd think would likely be used MathJax.

MathJax rendering in notebooks is (I think? I forget...) asynchronous though, which may be related.

**Comment by jfbercher** [comment]
Yes, certainly asynchronous since the literal maths `$\LaTeX$` are first shown in the toc and then rendered. The maths duplication issue seems similar to this [issue](https://github.com/mathjax/MathJax/issues/1619).

**Comment by PeterMinin** [comment]
@jcb91 Actually, after all, the conda-forge version of 0.2.2 *does* have an old version of toc2. If you download [jupyter_contrib_nbextensions-0.2.2-py27_0.tar.bz2](https://anaconda.org/conda-forge/jupyter_contrib_nbextensions/0.2.2/download/linux-64/jupyter_contrib_nbextensions-0.2.2-py27_0.tar.bz2) and look into `lib/python2.7/site-packages/jupyter_contrib_nbextensions/nbextensions/toc2/README.md`, the last line there starts with `- @jfbercher septemeber 21`.
Last time I was confused by the copy of the extension I had left in `.local/share/jupyter` from an installation via pip and `jupyter ... install --user`.
Also, thank you for the instructions.

**Comment by jcb91** [comment]
> Actually, after all, the conda-forge version of 0.2.2 does have an old version of toc2
> [...]
> the last line there starts with [- @jfbercher septemeber 21.]

Right, but that is the case also for 0.2.2. The 0.2.2 version _is_ old compared to the current repo master. We just haven't yet had a newer version on pip/anaconda that incorporates latest changes to toc2 (in fact, they seem to be the first PR merged _after_ 0.2.2 in bd65972332af73377c2fef7ec2586dfe3d28e28c). Essentially, I need to put 0.2.3 out :blush:

> Last time I was confused by the copy of the extension I had left in `.local/share/jupyter`

yeah, it gets confusing managing versions!

**Comment by PeterMinin** [comment]
Of course, the linear log view... I was judging by the fact that https://github.com/ipython-contrib/jupyter_contrib_nbextensions/commit/a61750af97d96e00a12bb012f66642d98b4a3bf5 is below https://github.com/ipython-contrib/jupyter_contrib_nbextensions/commit/0cdb295c9f667186f07d753bf260ef14d89a9a8b in GitHub's log...

**Comment by jcb91** [comment]
>  I was judging by the fact that a61750a is below 0cdb295 in GitHub's log...

not sure which log you mean here?

I've rolled a 0.2.3 release on pypi, but conda build will as usual be delated until the source dist is available on pypi.io

**Comment by PeterMinin** [comment]
Thank you.
I meant [the Commits page](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/commits/master), an analog of `git log`.

**Comment by jcb91** [comment]
> I meant [the Commits page](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/commits/master), an analog of `git log`.

Ah, neat, I'd never noticed that before! I've always used [SourceTree](https://www.sourcetreeapp.com/) / [GitKraken](https://www.google.co.uk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwiitMW8ld7QAhVL2RoKHU0MCG0QFggaMAA&url=https%3A%2F%2Fwww.gitkraken.com%2F&usg=AFQjCNFchzqmj3FicX4eeIhjyKyLcNgciA&sig2=j0oqknpTRQatscK_5BaCeg&bvm=bv.139782543,d.bGg) :)
Yeah, the log is kinda confusing in its ordering looked at like that...

**Comment by PeterMinin** [comment]
Same reason why I didn't expect this kind of magic, I always use at least `git log --graph`.

Ok, enough with the off-topic, sorry @jfbercher :)
